### PR TITLE
Memory efficiency blog for Valkey 8.0

### PR DESCRIPTION
**Pull request by hpatro**
### Description

Adding a blog describing the memory efficiency related contribution for Valkey 8
 
### Check List
- [x] Commits are signed per the DCO using `--signoff`

By submitting this pull request, I confirm that my contribution is made under the terms of the BSD-3-Clause License.

### CODE REVIEW

File: content/blog/2024-08-29-valkey-memory-efficiency-8-0.md [hunk 1]
```diff
@@ -0,0 +1,148 @@
++++
+# `title` is how your post will be listed and what will appear at the top of the post
```
Comment (stockholmux) [inline content/blog/2024-08-29-valkey-memory-efficiency-8-0.md:2]: you can remove the boiler plate comments :)

File: content/blog/2024-08-29-valkey-memory-efficiency-8-0.md [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+date= 2024-08-29 01:01:01
+# 'description' is what is shown as a snippet/summary in various contexts.
+# You can make this the first few lines of the post or (better) a hook for readers.
+# Aim for 2 short sentences.
+description= "Learn about the new memory efficiency improvements in Valkey 8 which reduces memory overhead, allows more data to be stored in the same amount of memory."
+# 'authors' are the folks who wrote or contributed to the post.
+# Each author corresponds to a biography file (more info later in this document)
+authors= [ "hpatro"]
++++
+
```
Comment (stockholmux) [inline content/blog/2024-08-29-valkey-memory-efficiency-8-0.md:17]: One line per sentence plz.

File: content/blog/2024-08-29-valkey-memory-efficiency-8-0.md [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+Valkey 8.0 GA is around the corner and one of the theme is increasing overall memory efficiency. Memory overhead reduction has the obvious effect of better resource utilization, but also impacts performance. By minimizing unnecessary memory consumption, you can store more data with the same hardware resources and improve overall system responsiveness. This post is going to give a overview into how Valkey internally manages the data and its memory overhead. Additionally, it talks about the two major improvements for Valkey 8.0 that improves the overall memory efficiency.
```
Comment (stockholmux) [inline content/blog/2024-08-29-valkey-memory-efficiency-8-0.md:18]: `one of the theme is increasing` -> `one of the themes is increasing`

Comment (stockholmux) [reply content/blog/2024-08-29-valkey-memory-efficiency-8-0.md:18]: `This post is going to give a overview` -> `This post is going to give an overview`

File: content/blog/2024-08-29-valkey-memory-efficiency-8-0.md [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+![Figure 2 Dictionary bucket pointing to a dictionary entry](/assets/media/pictures/dictionary_bucket_and_entry_overview.png)
+
+In cluster mode, Valkey uses a concept called [hash slots](https://valkey.io/topics/cluster-tutorial/) to shard data. There are 16384 hash slots in cluster, and to compute the hash slot for a given key, the server computes the CRC16 of the key modulo 16384. Keys are distributed on basis of these slots assigned to each of the primary. The server needs to maintain additional metadata for bookkeeping i.e. slot-to-key mapping to move a slot from one primary to another. In order to maintain the slot to key mapping, two additional pointers `slot-prev` and `slot-next` (Figure 3)  are stored as metadata in each dictionary entry forming a double linked list of all keys belonging to a given slot. This further increases the overhead by 16 bytes per dictionary entry i.e. total 40 bytes.
```
Comment (stockholmux) [inline content/blog/2024-08-29-valkey-memory-efficiency-8-0.md:30]: `There are 16384 hash slots in cluster` -> `There are 16,384 hash slots in a cluster`

Comment (stockholmux) [reply content/blog/2024-08-29-valkey-memory-efficiency-8-0.md:30]: `key modulo 16384` -> `key modulo 16,384`

File: content/blog/2024-08-29-valkey-memory-efficiency-8-0.md [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+![Figure 3 Dictionary in cluster mode (Valkey 7.2) with multiple key value pair](/assets/media/pictures/dictionary_in_cluster_mode_7.2.png)
+
+## Improvements
+
+### Optimization 1 - [Dictionary per slot](https://github.com/redis/redis/pull/11695)
+
+The first optimization is we have introduced a dictionary per slot (16,384 of them in total), where each dictionary stores data for a given slot. With this simplification, the cost of maintaining additional metadata for the mapping of slot to key is no longer required in Valkey 8. To iterate over all the keys in a given slot, the engine simply finds out the dictionary for a given slot and traverse all the entries in it. This reduces the memory usage per dictionary entry by 16 bytes with a small memory overhead around 1 MB per node. As cluster mode is generally used for storing large amount of keys, avoiding the additional overhead per key allows users to store more number of keys in the same amount of memory.
```
Comment (stockholmux) [inline content/blog/2024-08-29-valkey-memory-efficiency-8-0.md:38]: `The first optimization is we have introduced a dictionary per slot` doesn't quite parse for me.
Maybe `The first optimization is a dictionary per slot`?

File: content/blog/2024-08-29-valkey-memory-efficiency-8-0.md [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+![Figure 4 Dictionary in cluster mode (Valkey 8.0) with multiple key value pair](/assets/media/pictures/dictionary_in_ cluster_mode_8.0.png)
```
Comment (stockholmux) [inline content/blog/2024-08-29-valkey-memory-efficiency-8-0.md:40]: This doesn't render... here is how it looks in situ:

Screenshot 2024-08-30 at 1 21 30 PM

File: content/blog/2024-08-29-valkey-memory-efficiency-8-0.md [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+Few of the interesting challenges that comes up with the above improvements are supporting existing use cases which were optimized with a single dictionary for the entire keyspace. The usecases are:
```
Comment (stockholmux) [inline content/blog/2024-08-29-valkey-memory-efficiency-8-0.md:42]: `Few of the interesting challenges` -> `A few of the interesting challenges`

File: content/blog/2024-08-29-valkey-memory-efficiency-8-0.md [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+* [Iterating the entire keyspace](https://valkey.io/commands/scan/) - Command like `SCAN` to iterate over the entire keyspace.
+* [Random key for eviction](https://valkey.io/topics/lru-cache/) - The server does random sampling of the keyspace to find ideal candidate for eviction.
+* [Finding a random key](https://valkey.io/commands/randomkey/) - Command like `RANDOMKEY` retrieves a random key from the database.
+
+
+It is quite possible that some slots contains more keys than others and some could be also empty which could lead the engine to traverse through the all the empty dictionary for a full scan i.e. [O(N)](https://en.wikipedia.org/wiki/Big_O_notation) operation if approached naively. When picking a random key for eviction or running `RANDOMKEY` operation, the server needs to pick a non-empty slot in a performant manner. These cases require a data structure which provides the following functionality:
```
Comment (stockholmux) [inline content/blog/2024-08-29-valkey-memory-efficiency-8-0.md:49]: `could lead the engine to traverse through the all the empty dictionary` -> `could lead the engine to traverse through the all empty dictionary` (I think that's what you meant).

File: content/blog/2024-08-29-valkey-memory-efficiency-8-0.md [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+
+1. Modify value for a given index (slot) - If a key gets added or removed, increment or decrement the value for that given slot by 1 respectively.
+2. Cumulative frequency until each index (slot) - For a given number representing a key between 1 and the total number of keys, return the index (slot) which covers the particular key.
+
+If approached naively, the former and latter operation would take O(1) and O(N) respectively. However, we want to minimize the latter operation’s time complexity and avoid regression in the operations stated above. Hence, a [binary indexed tree (BIT) or fenwick tree](https://www.topcoder.com/thrive/articles/Binary%20Indexed%20Trees) which provides the above functionality with a minimal memory overhead (~1 MB per node) and the time complexity is also bounded to O(M log N) for both the operations where M = number of modification(s) and N = number of slots. This enables skipping over empty slot efficiently while iterating over the keyspace as well as finding a slot for a given key index in logarithmic time via binary search over the cummulative sum maintained by the BIT.
```
Comment (stockholmux) [inline content/blog/2024-08-29-valkey-memory-efficiency-8-0.md:54]: `This enables skipping over empty slot efficiently` -> `This enables skipping over empty slots efficiently`

File: content/blog/2024-08-29-valkey-memory-efficiency-8-0.md [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+
+Overall with this new approach, the benefits are: 
+
+1. Removes additional memory overhead in cluster mode: Get rid of two pointers (16 bytes) per key to keep the mapping of slot to keys.
+2. Rehashing operation is CPU intensive: By default, limited number of buckets are allocated in a dictionary and it expands/shrinks dyanmically based on usage. While undergoing rehashing, all the data needs to be moved from an old dictionary to a new dictionary. With Valkey 7.2, a global dictionary being shared across all the slots, all the keys gets stored under a single dictionary and each time the fill factor (number of keys / number of buckets) goes above 1.5, the dictionary needs to a large dictionary (multiple of 2) and move large amount of keys. As this operation is performed on the fly, it causes increase in latency for regular command operations while it’s ongoing. With this optimization, the impact of rehashing is localized to the specific dictionary undergoing the process and only subset of key needs to be moved.
```
Comment (stockholmux) [inline content/blog/2024-08-29-valkey-memory-efficiency-8-0.md:59]: The verb is missing here: 
`the dictionary needs to a large dictionary` -> `the dictionary needs to move to a larger dictionary` (I think...)

File: content/blog/2024-08-29-valkey-memory-efficiency-8-0.md [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+
+### Optimization 2 - [Key embedding into dictionary entry](https://github.com/valkey-io/valkey/pull/541)
+
+After the dictionary per slot change, the memory layout of dictionary entry in cluster mode is the following, there are three pointers (key, value, and next).  The key pointer points to a SDS([simple dynamic string](https://github.com/antirez/sds/blob/master/README.md)) which contains the actual key data. As key is immutable, without bringing in much complexity, it can be embedded into the dictionary entry which has the same lifetime as the former.
```
Comment (stockholmux) [inline content/blog/2024-08-29-valkey-memory-efficiency-8-0.md:63]: ` SDS([simple dynamic string](https://github.com/antirez/sds/blob/master/README.md))` -> ` SDS ([simple dynamic string](https://github.com/antirez/sds/blob/master/README.md))` (needs a space next to the open paren)

File: content/blog/2024-08-29-valkey-memory-efficiency-8-0.md [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+
+![Figure 5 Key data storage in 7.2 (left) and 8.0 (right)](static/assets/media/pictures/key_embedding.png)
```
Comment (stockholmux) [inline content/blog/2024-08-29-valkey-memory-efficiency-8-0.md:65]: This doesn't render properly

Screenshot 2024-08-30 at 1 32 00 PM

File: content/blog/2024-08-29-valkey-memory-efficiency-8-0.md [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+(integer) 6318941
+127.0.0.1:6381> INFO MEMORY # command to retrieve statistics about memory usage
+# Memory
+used_memory:577300952
+used_memory_human:550.56M
+```
+
+#### Overall Improvement
+
+![Figure 6 Overall memory usage with benchmark data](static/assets/media/pictures/memory_usage comparison.png)
```
Comment (stockholmux) [inline content/blog/2024-08-29-valkey-memory-efficiency-8-0.md:129]: Doesn't render:

Screenshot 2024-08-30 at 1 34 30 PM

File: content/authors/hpatro.md [hunk 14]
```diff
@@ -0,0 +1,11 @@
+---
+title: Harkrishn Patro
+extra:
+    photo: '/assets/media/authors/hpatro.png'
```
Comment (stockholmux) [inline content/authors/hpatro.md:4]: I prefer a real photo, but if you don't have one you like, no problem.

File: content/authors/hpatro.md [hunk 15]
```diff
@@ -0,0 +1,11 @@
...
+    github: hpatro
+---
+
+Harkrishn Patro is a Software Engineer at AWS.
+
+Harkrishn is passionate about distributed systems, databases, anything related to data.
```
Comment (stockholmux) [inline content/authors/hpatro.md:10]: Any thing you like outside of work? Badminton, surfing, making hats, etc.?

**Review by stockholmux — CHANGES_REQUESTED** [review]
Mostly minor grammar things, a few image issues, but overall a nice writeup.

**Comment by stockholmux** [comment]
Given the holiday weekend (and Open Source Summit EU), I'd like to target to publish this Sept 3 or 4.

File: content/authors/hpatro.md (hunk 15 above)
Comment (hpatro) [reply content/authors/hpatro.md:10]: Somehow you guess my favorite sport to play. Will add some :)

File: content/authors/hpatro.md (hunk 14 above)
Comment (hpatro) [reply content/authors/hpatro.md:4]: Couldn't decide which photo to add and copied a default one. :D

File: content/blog/2024-08-29-valkey-memory-efficiency-8-0.md (hunk 8 above)
Comment (hpatro) [reply content/blog/2024-08-29-valkey-memory-efficiency-8-0.md:39]: I meant `could lead the engine to traverse through all the empty dictionary`

**Comment by hpatro** [comment]
@stockholmux Thanks a lot for your feedback. Updated the PR, PTAL.

File: static/assets/media/pictures/memory_usage_comparison.png
Comment (madolson) [inline static/assets/media/pictures/memory_usage_comparison.png:1]: can this graph start with 0 on the y axis? The fact it is floating bothers me.

Comment (hpatro) [reply static/assets/media/pictures/memory_usage_comparison.png:1]: Ha ha, agreed it's an eye sore. excalidraw does that by default. Fixed.

**Review by madolson — COMMENTED** [review]
Mostly looks good to me now. Some minor wording and phrasing suggestions.

File: content/blog/2024-08-29-valkey-memory-efficiency-8-0.md [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+
+### Optimization 2 - [Key embedding into dictionary entry](https://github.com/valkey-io/valkey/pull/541)
+
+After the dictionary per slot change, the memory layout of dictionary entry in cluster mode is the following, there are three pointers (key, value, and next).  The key pointer points to a SDS ([simple dynamic string](https://github.com/antirez/sds/blob/master/README.md)) which contains the actual key data. As key is immutable, without bringing in much complexity, it can be embedded into the dictionary entry which has the same lifetime as the former.
```
[the remark this answers is not in the document]
Comment (hpatro) [reply content/blog/2024-08-29-valkey-memory-efficiency-8-0.md:53]: `RENAME` command doesn't modify it in place. It creates a new key and removes the old one, so the key data is immutable.

File: content/blog/2024-08-29-valkey-memory-efficiency-8-0.md [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+
+Overall with this new approach, the benefits are: 
+
+1. Removes additional memory overhead in cluster mode: Get rid of two pointers (16 bytes) per key to keep the mapping of slot to keys.
+2. Rehashing operation is CPU intensive: By default, limited number of buckets are allocated in a dictionary and it expands/shrinks dyanmically based on usage. While undergoing rehashing, all the data needs to be moved from an old dictionary to a new dictionary. With Valkey 7.2, a global dictionary being shared across all the slots, all the keys gets stored under a single dictionary and each time the fill factor (number of keys / number of buckets) goes above 1.5, the dictionary needs to move to a larger dictionary (multiple of 2) and move large amount of keys. As this operation is performed on the fly, it causes increase in latency for regular command operations while it’s ongoing. With this optimization, the impact of rehashing is localized to the specific dictionary undergoing the process and only subset of key needs to be moved.
```
[the remark this answers is not in the document]
Comment (hpatro) [reply content/blog/2024-08-29-valkey-memory-efficiency-8-0.md:49]: Yeah, should I place it before the `Overall with this new approach` section ?

**Review by madolson — APPROVED** [review]
LGTM

File: content/blog/2024-08-29-valkey-memory-efficiency-8-0.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (stockholmux) [reply content/blog/2024-08-29-valkey-memory-efficiency-8-0.md]: @hpatro TIL about `RENAME`.

File: content/blog/2024-08-29-valkey-memory-efficiency-8-0.md [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+
+Another interesting side effect is on the rehashing operation. Rehashing is CPU intensive. By default, a limited number of buckets are allocated in a dictionary and it expands/shrinks dynamically based on usage. While undergoing rehashing, all the data needs to be moved from an old dictionary to a new dictionary. With Valkey 7.2, a global dictionary being shared across all the slots, all the keys get stored under a single dictionary and each time the fill factor (number of keys / number of buckets) goes above 1, the dictionary needs to move to a larger dictionary (multiple of 2) and move a large amount of keys. As this operation is performed on the fly, it causes an increase in latency for regular command operations while it's ongoing. With the per-slot dictionary optimization, the impact of rehashing is localized to the specific dictionary undergoing the process and only a subset of key needs to be moved.
```
Comment (stockholmux) [inline content/blog/2024-08-29-valkey-memory-efficiency-8-0.md:46]: `only a subset of key needs` -> `only a subset of keys needs` (make keys plural)

File: content/blog/2024-08-29-valkey-memory-efficiency-8-0.md [hunk 19]
```diff
@@ -0,0 +1,140 @@
++++
+title= "Storing more with less: Memory Efficiency in Valkey 8"
+date= 2024-08-29 01:01:01
```
Comment (stockholmux) [inline content/blog/2024-08-29-valkey-memory-efficiency-8-0.md:3]: Update to `2024-09-04 01:01:01`

**Review by stockholmux — CHANGES_REQUESTED** [review]
Nice work @hpatro! I found one small issue and we need to update the date to post (so it isn't in the past) but we can merge after that!

**Review by stockholmux — APPROVED** [review]
LGTM

**Pull request merged**